lib/rules/selector-max-type/__tests__/index.mjs

```diff
@@ -138,6 +129,24 @@ testRule({
 			endLine: 1,
 			endColumn: 60,
 		},
+		{
+			code: '.foo { &-bar {} }',
+			description: '&-bar is an invalid selector, types must precede `&`',
+			message: messages.expected('&-bar', 0),
+			line: 1,
+			column: 8,
+			endLine: 1,
+			endColumn: 13,
+		},
+		{
+			code: '.foo { &__bar {} }',
+			description: '&__bar is an invalid selector, types must precede `&`',
+			message: messages.expected('&__bar', 0),
+			line: 1,
+			column: 8,
+			endLine: 1,
+			endColumn: 14,
+		},
```

Comment:
This is a breaking change.

If people use this rule while writing sass style nesting they will get errors after updating. Given the popularity of sass style nesting it will definitely be noticed when we change this.

Reply:
Maybe add a specific `if` to detect the pattern and `console.warn` a hint?

Reply:
Or should we add detection for this in `isStandardSyntaxSelector`?

`&__bar` and `&-bar` are never standard selectors, they can only be sass, less, ...
So I think `isStandardSyntaxSelector` should return false in this case.

I think this might be the least intrusive way forward.

Reply:
> are _never_ standard selectors

If true, it ought to be fixed in a separate commit.
IIRC this util is being exported so it might have consequences.
We shouldn't care though, utils ought to match their names.

Reply:
I agree with adding detection to `isStandardSyntaxSelector` by another pull request, as @Mouvedia suggested.

Reply:
I think this PR is the right approach we should take 👍🏼 
But one big concern is a breaking change for non-standard syntaxes, e.g., SCSS, as you considered.

My current guess is that we have to bring the breaking change in the next major version, especially for the SCSS users (I guess it's a big community).

Reply:
Yup, definitely prefer saving this for a next major.

I do find these changes very informative.
There are a lot of past choices encoded in these rules and refactoring them is a good way to get a better feel for the complexity.

Maybe we can setup a long lived branch for selector rule refactors?
Its main purpose would be to explore the coming changes so that possible issues can be encountered early. Either we then pick specific parts of that branch or we delete it entirely and start over manually, but with a fresh and comprehensive understanding of the work that should be done.

Reply:
Maybe there are others selector-* rules that use `postcss-resolve-nested-selector` that don't have problematic test cases. If so, the regression won't be considered as severe because it wasn't acknowledged though the uses of test cases. It's a technicality but that would pass as sub-major change in my book.

Reply:
I agree with the refactoring using the nice package `"@csstools/selector-resolve-nested` unless it introduces breaking changes. 👍🏼 

I believe a new rule might need this package for spec-compliant selector parsing.

Reply:
> I agree with the refactoring using the nice package "@csstools/selector-resolve-nested unless it introduces breaking changes.

I am a bit confused now :)
I think that any refactor using this package will have some amount of breaking changes and should be saved for a next major.

I think we have these options:
1. do not use this package at this time, save it for a next major. We close these PR's and focus purely on the issue of inaccurate end positions which might also have a different solution.
2. Start work early on the next major and start refactoring all selector rules to use this package. How exactly we work on the next major, this early, is a different topic :)
3. A mix of 1 and 2 where we fix inaccurate end positions in a way that facilitates a future switch to this package. This also implies having a good understanding of this package and how it will integrate in the future.

Option 1 might not find a solution and it can not solve issues with nesting. But it could be a non-breaking step forward.

Notes :
- I am not attached to these changes, the proposed package, ... These can all be discarded if they are not the right way forward.
- I am fine with shelving this work for a few months. Closing these PR's and reviving the effort later.

Reply:
Thank you for the detailed explanation. I misunderstood there was room to refactor via `@csstools/selector-resolve-nested` somewhere in the code base. 👍🏼 

We just released the current major version (16.0.0) a few months ago (Dec 2023), so I predict the next major release will be several months later. I guess staring the work for the next major version is a bit too early at this point.

Of course, I welcome non-breaking changes. Also, since end-position bugs are a bit less critical, we could address other works, e.g. considering the migration path from `postcss-resolve-nested-selector` to `@csstools/selector-resolve-nested`.

Reply:
For the record, I vote 3.

Reply:
I have a vague idea of what 3 looks like.
I will try to prepare a PR with that shortly.

Then this is less abstract :)

Reply:
see : https://github.com/stylelint/stylelint/pull/7518